### PR TITLE
added login,logout,userapiview

### PR DESCRIPTION
**Pull request by osadolortech**

### CODE REVIEW

File: tweetapi/login.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+
+class Login(CreateAPIView):
+    serializer_class = UserSerializer
+    def post(self, request, *args, **kwargs):
+        username = request.data.get("username")
+        password = request.data.get("password")
+
+        if not username or not password:
+            return Response({"error":"please fill all fields"}, status=status.HTTP_400_BAD_REQUEST)
```
Comment (iamr0b0tx) [inline tweetapi/login.py]: You can use a serializer here

File: tweetapi/login.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        check_user = User.objects.filter(username=username)
+        if check_user == False:
+            return Response({"error":"username does not exsit"})
```
Comment (iamr0b0tx) [inline tweetapi/login.py]: why do this when you are going to check if user is none. you are making two db calls for the same thing

File: tweetapi/login.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        if user is not None:
+            login(request,user)
+            payload = {
+                "id":user.id,
+                "exp":datetime.timedelta(minutes=2) + datetime.datetime.utcnow(),
+                "iat":datetime.datetime.utcnow()
+            }
+            payload2 = {
+                "id":user.id,
+                "exp":datetime.timedelta(days=3) + datetime.datetime.utcnow(),
```
Comment (iamr0b0tx) [inline tweetapi/login.py]: try not to hardcode numbers. add it to settings

File: tweetapi/login.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        password = request.data.get("password")
+
+        if not username or not password:
+            return Response({"error":"please fill all fields"}, status=status.HTTP_400_BAD_REQUEST)
+        check_user = User.objects.filter(username=username)
+        if check_user == False:
+            return Response({"error":"username does not exsit"})
+        user = authenticate(username=username,password=password)
+        if user is not None:
+            login(request,user)
```
Comment (iamr0b0tx) [inline tweetapi/login.py]: why use `login` it is a different authentication scheme

**Review by iamr0b0tx — CHANGES_REQUESTED** [review]

File: tweetapi/login.py (hunk 1 above)
Comment (osadolortech) [reply tweetapi/login.py]: Alright sir, notes will make changes

File: tweetapi/login.py (hunk 2 above)
Comment (osadolortech) [reply tweetapi/login.py]: I didn't actually think about it this way, I tot I was trying to make all errors was accounted for....I just realized I was over doing it. Thank you sir

File: tweetapi/login.py (hunk 3 above)
Comment (osadolortech) [reply tweetapi/login.py]: Alright sir

File: tweetapi/login.py (hunk 2 above)
Comment (osadolortech) [reply tweetapi/login.py]: fixed

File: tweetapi/login.py (hunk 3 above)
Comment (osadolortech) [reply tweetapi/login.py]: fixed

File: tweetapi/login.py (hunk 4 above)
Comment (osadolortech) [reply tweetapi/login.py]: fixed sir

**Review by iamr0b0tx — APPROVED** [review]

File: tweetapi/login.py (hunk 1 above)
Comment (iamr0b0tx) [reply tweetapi/login.py]: ?

Comment (osadolortech) [reply tweetapi/login.py]: this is outdated sir made those changes already

File: tweetapi/login.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+from rest_framework.response import Response
+from django.conf import settings
+
+
+
+class Login(CreateAPIView):
+    serializer_class = UserSerializer
+    def post(self, request, *args, **kwargs):
+        username = request.data.get("username")
+        password = request.data.get("password")
```
Comment (iamr0b0tx) [inline tweetapi/login.py]: how exactly have you used a serializer?

**Review by iamr0b0tx — CHANGES_REQUESTED** [review]

**Comment by osadolortech** [comment]
I  thought you meant the serializer_class will do that for me that I don’t
need all of the validations I wrote, tried using the serializer and it keep
saying user already exist and it won’t let me logging to generate token.
How am I supposed to use the serializer sir.

**Comment by iamr0b0tx** [comment]
I will let you figure that one out yourself

**Comment by osadolortech** [comment]
> I will let you figure that one out yourself

figured it out sir, made a serializer for userloginserializer instead of using my userserilaizer

File: tweetapi/login.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+from rest_framework import status
+from rest_framework.response import Response
+from django.conf import settings
+
+
+
+class Login(CreateAPIView):
+    serializer_class = UserloginSerializer
+    def post(self, request,format=None):
+        serializer = UserloginSerializer(data=request.data)
```
Comment (iamr0b0tx) [inline tweetapi/login.py:19]: nice work. checkout `get_serializer_class`

File: tweetapi/login.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        if serializer.is_valid(raise_exception=True):
```
Comment (iamr0b0tx) [inline tweetapi/login.py]: this is not how to use `is_valid` and `raise_exception`

**Review by iamr0b0tx — CHANGES_REQUESTED** [review]

File: tweetapi/login.py (hunk 7 above)
Comment (osadolortech) [reply tweetapi/login.py]: i was trying to use is_valid() method to perform validation for each field, raise_exception=True flag, i was looking to override the style of validation error responses, how can i use it, and am i to use the is_valid() without the raise_exception and hows the best way to use it sir

Comment (iamr0b0tx) [reply tweetapi/login.py]: you can't check is valid and still raise exceptions at the same time. pick one

Comment (iamr0b0tx) [reply tweetapi/login.py]: `serializer.is_valid(raise_exception=True)`

or

`if serializer.is_valid():`

**Comment by osadolortech** [comment]
Ok sir……

File: tweetapi/login.py (hunk 7 above)
Comment (osadolortech) [reply tweetapi/login.py]: fixed sir

**Review by iamr0b0tx — APPROVED** [review]

**Pull request merged**